### PR TITLE
Custom View Optimization

### PR DESCRIPTION
**Pull request by dandc87**
This will be a breaking change. the way the view reacts to layout parameters is improved. The functionality is similar to how `adjustViewBounds` works with ImageView, where the view will try and take up as much space as necessary, but we base it on tile size instead of an aspect ratio. The exception being that if a `tileSize` is set, that will override everything and set the view to that size.

This also better, but still incomplete, support for editing mode. You can now see a preview of a month instead of just whitespace.

And finally, `MonthPagerAdapter` was moved to it's own file, since `MaterialCalendarView` was getting a little big.

Issues this should close: #84, #44, #87, #46

### CODE REVIEW

**Comment by dandc87** [comment]
I've uploaded a version of the library to [Sonatype's `snapshots` repository](https://oss.sonatype.org/content/repositories/snapshots/com/prolificinteractive/material-calendarview/) for testing.

File: library/src/main/java/com/prolificinteractive/materialcalendarview/MaterialCalendarView.java [hunk 1]
```diff
@@ -242,31 +246,27 @@ public void transformPage(View page, float position) {
 
         currentMonth = CalendarDay.today();
         setCurrentDate(currentMonth);
+
+        if(isInEditMode()) {
+            removeView(pager);
+            MonthView monthView = new MonthView(context, currentMonth, getFirstDayOfWeek());
+            monthView.setSelectionColor(getSelectionColor());
+//            monthView.setWeekDayFormatter(weekDayFormatter);
+//            monthView.setDayFormatter(dayFormatter);
```
Comment (ireneduke) [inline library/src/main/java/com/prolificinteractive/materialcalendarview/MaterialCalendarView.java]: What are these commented out lines doing here?

File: library/src/main/java/com/prolificinteractive/materialcalendarview/MaterialCalendarView.java [hunk 2]
```diff
@@ -752,13 +747,13 @@ protected void onRestoreInstanceState(Parcelable state) {
     }
 
     @Override
-    protected void dispatchSaveInstanceState(SparseArray<Parcelable> container) {
+    protected void dispatchSaveInstanceState(@NonNull SparseArray<Parcelable> container) {
         //super.dispatchSaveInstanceState(container);
```
Comment (ireneduke) [inline library/src/main/java/com/prolificinteractive/materialcalendarview/MaterialCalendarView.java]: Why is this line here?

File: library/src/main/java/com/prolificinteractive/materialcalendarview/MaterialCalendarView.java [hunk 3]
```diff
@@ -752,13 +747,13 @@ protected void onRestoreInstanceState(Parcelable state) {
...
         super.dispatchFreezeSelfOnly(container);
     }
 
     @Override
-    protected void dispatchRestoreInstanceState(SparseArray<Parcelable> container) {
+    protected void dispatchRestoreInstanceState(@NonNull SparseArray<Parcelable> container) {
         //super.dispatchRestoreInstanceState(container);
```
Comment (ireneduke) [inline library/src/main/java/com/prolificinteractive/materialcalendarview/MaterialCalendarView.java]: Why is this line here?

File: library/src/main/java/com/prolificinteractive/materialcalendarview/MonthPagerAdapter.java [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+import java.util.LinkedList;
+import java.util.List;
+
+/**
+ * Pager adapter backing the calendar view
+ */
+class MonthPagerAdapter extends PagerAdapter {
+
+    private final LinkedList<MonthView> currentViews;
+    private final ArrayList<CalendarDay> months;
```
Comment (ireneduke) [inline library/src/main/java/com/prolificinteractive/materialcalendarview/MonthPagerAdapter.java]: This is a little confusing. Does the CalendarDay object represent a month?

Comment (dandc87) [inline library/src/main/java/com/prolificinteractive/materialcalendarview/MonthPagerAdapter.java]: It's a list of the first days of months. I'm working on a more efficient way of paging months in the [aek/month-math](https://github.com/prolificinteractive/material-calendarview/compare/aek/custom-views...aek/month-math) branch, so this should become moot.

**Comment by dandc87** [comment]
@ireneduke cleaned stuff up based on your feedback

**Comment by ireneduke** [comment]
@dandc87 Should the readme be updated since this a breaking change?

**Comment by dandc87** [comment]
@ireneduke since the readme is documentation for the current release, and I don't plan on immediately releasing a new version with these changes, I'm forced to update the README asynchronously from the actual code in master :cry:

**Comment by ireneduke** [comment]
@dandc87 Maybe this should be merged to a develop branch if you don't plan on releasing it immediately?

**Comment by dandc87** [comment]
Closing in favor of #104
